lib/stdlib/src/supervisor.erl

```diff
@@ -320,31 +321,30 @@ init_dynamic(_State, StartSpec) ->
 
 %%-----------------------------------------------------------------
 %% Func: start_children/2
-%% Args: Children = [child_rec()] in start order
+%% Args: Children = children() % ids in start order
 %%       SupName = {local, atom()} | {global, atom()} | {pid(), Mod}
-%% Purpose: Start all children.  The new list contains #child's
+%% Purpose: Start all children.  The new map contains #child's
 %%          with pids.
 %% Returns: {ok, NChildren} | {error, NChildren, Reason}
-%%          NChildren = [child_rec()] in termination order (reversed
-%%                        start order)
+%%          NChildren = children() % ids in termination order
+%%                                   (reversed start order)
 %%-----------------------------------------------------------------
-start_children(Children, SupName) -> start_children(Children, [], SupName).
-
-start_children([Child|Chs], NChildren, SupName) ->
-    case do_start_child(SupName, Child) of
-	{ok, undefined} when ?is_temporary(Child) ->
-	    start_children(Chs, NChildren, SupName);
-	{ok, Pid} ->
-	    start_children(Chs, [Child#child{pid = Pid}|NChildren], SupName);
-	{ok, Pid, _Extra} ->
-	    start_children(Chs, [Child#child{pid = Pid}|NChildren], SupName);
-	{error, Reason} ->
-	    report_error(start_error, Reason, Child, SupName),
-	    {error, lists:reverse(Chs) ++ [Child | NChildren],
-	     {failed_to_start_child,Child#child.id,Reason}}
-    end;
-start_children([], NChildren, _SupName) ->
-    {ok, NChildren}.
+start_children(Children, SupName) ->
+    Fun =
+        fun(Id,Child) ->
+                case do_start_child(SupName, Child) of
+                    {ok, undefined} when ?is_temporary(Child) ->
+                        remove;
+                    {ok, Pid} ->
+                        {update,Child#child{pid = Pid}};
+                    {ok, Pid, _Extra} ->
+                        {update,Child#child{pid = Pid}};
+                    {error, Reason} ->
+                        report_error(start_error, Reason, Child, SupName),
+                        {abort,{failed_to_start_child,Id,Reason}}
+                end
+        end,
+    ch_map(Fun,Children).
```

Comment:
I think it would be easier to understand if you changed Fun to Start and ch_map to children_map so that last
line would read children_map(Start, Children).

Reply:
Yes, I agree!